presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java

```diff
@@ -95,9 +106,28 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
                 e.getAuthenticateHeader().ifPresent(authenticateHeaders::add);
                 continue;
             }
-
             // authentication succeeded
-            nextFilter.doFilter(withPrincipal(request, principal), response);
+            CustomHttpServletRequestWrapper wrappedRequest = new CustomHttpServletRequestWrapper(request);
+            Map<String, String> extraHeadersMap = new HashMap<>();
+
+            for (ClientRequestFilter requestFilter : clientRequestFilterManager.getClientRequestFilters()) {
+                boolean headersPresent = requestFilter.getHeaderNames().stream()
+                        .allMatch(headerName -> request.getHeader(headerName) != null);
+
+                if (!headersPresent) {
+                    Optional<Map<String, String>> extraHeaderValueMap = requestFilter.getExtraHeaders(principal);
+
+                    extraHeaderValueMap.ifPresent(map -> {
+                        for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                            if (request.getHeader(extraHeaderEntry.getKey()) == null) {
+                                extraHeadersMap.putIfAbsent(extraHeaderEntry.getKey(), extraHeaderEntry.getValue());
```

Comment:
I don't think we should silently not put the header if there's an existing header.  I think we should consider the following:

1. The SPI contract mandates that any potential header is known upfront.
2. The runtime will validate that the headers that are requested to be modified match with what is actually returned.
3. There is a blocklist of headers that the plugin should not modify (such as query ID).
4. The runtime will validate that all registered filters only return a disjoint set of headers being added.  If there is a conflict, this is a fatal error that will require an administrator to fix (but un-registering one of these plugins).
5. If there is a conflict here, then an exception is thrown.

Reply:
@tdcmeehan 
I have gone through your comments and added a few validations in the current implementation. I can verify that with you. However, I have to know about the headers in the blocklist that should not be modified.
Could you help?

Reply:
Let's start with the query ID, trace ID and transaction ID in PrestoHeaders.java.

Reply:
I have incorporated the changes mentioned in your comment. Could you please review them and share your suggestions?